### PR TITLE
Digest on RelNode and RexCall

### PR DESCRIPTION
**Pull request by hsyuan**
See discussion in CALCITE-3786.

### CODE REVIEW

**Comment by XuQianJin-Stars** [comment]
Features +1，It seems a bit broken here(RelDigest)?

**Comment by hsyuan** [comment]
Comparing with version 1.23.0, it is.
Comparing with current master, it is a less broken change.

**Comment by jinxing64** [comment]
+1 on this and thanks a lot @hsyuan, RelDigest is much concise.
Just one question, if the `RelDigest` is regarded as internal, why not it's not an inner class in AbstractRelNode ?

File: core/src/main/java/org/apache/calcite/plan/RelOptNode.java [hunk 1]
```diff
@@ -45,9 +47,20 @@
    * <p>If you want a descriptive string which contains the identity, call
    * {@link Object#toString()}, which always returns "rel#{id}:{digest}".
    *
+   * @return Digest string of this {@code RelNode}
+   */
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/RelOptNode.java:51]: Thanks a lot to keep this `String getDigest`, which is heavily relied on in our internal code base.

**Comment by danny0405** [comment]
> +1 on this and thanks a lot @hsyuan, RelDigest is much concise.
> Just one question, if the `RelDigest` is regarded as internal, why not it's not an inner class in AbstractRelNode ?

If possible, please get involve in the discussion in the CALCITE-3786 first.

**Comment by jinxing64** [comment]
@danny0405 
What you mean by "get involve in the discussion in the CALCITE-3786 first";
I checked this PR and give my approve. From my point it's indeed much more concise than `Digest` in the current master. And it also gives enough consideration on backward compatibility.

**Comment by yanlin-Lynn** [comment]
Actually, I think the implementation here do not have much difference with the master, just with moving the Digest related logic to InnerRelDigest.
+1 for the concise and backward compatibility.

**Comment by hsyuan** [comment]
> Just one question, if the RelDigest is regarded as internal, why not it's not an inner class in AbstractRelNode ?

Planner also use it, they are not in the same package. So I think we don't have choice.

File: core/src/main/java/org/apache/calcite/plan/RelDigest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+@API(since = "1.24", status = API.Status.INTERNAL)
+public interface RelDigest {
+  /**
+   * Reset state, possibly cache of hash code.
+   */
+  void clear();
+
+  /**
+   * Returns the relnode that this digest is associated with.
+   */
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/RelDigest.java:40]: Any benchmark for the improvement?

File: core/src/main/java/org/apache/calcite/rex/RexSubQuery.java [hunk 3]
```diff
@@ -139,11 +138,14 @@ public RexSubQuery clone(RelNode rel) {
 
   @Override public boolean equals(Object obj) {
     return obj == this
-        || obj instanceof RexCall
+        || obj instanceof RexSubQuery
         && toString().equals(obj.toString());
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rex/RexSubQuery.java:142]: Nice catch

**Comment by danny0405** [comment]
> @danny0405
> What you mean by "get involve in the discussion in the [CALCITE-3786](https://issues.apache.org/jira/browse/CALCITE-3786) first";
> I checked this PR and give my approve. From my point it's indeed much more concise than `Digest` in the current master. And it also gives enough consideration on backward compatibility.

I have replied in the issue, please also move there.

**Pull request merged**